assets/js/interactivity/store.ts

```diff
@@ -0,0 +1,281 @@
+				return scope.getters.get( getter ).value;
+			}
+		}
+
+		const result = Reflect.get( target, key, receiver );
+
+		// Check if the property is a generator.
+		if ( result?.constructor?.name === 'GeneratorFunction' ) {
+			return async ( ...args: unknown[] ) => {
+				const scope = getScope();
+				const gen: Generator< any > = result( ...args );
+
+				let value: any;
+				let it: IteratorResult< any >;
+
+				while ( true ) {
+					setNamespace( ns );
+					setScope( scope );
+					try {
+						it = gen.next( value );
+					} finally {
+						resetScope();
+						resetNamespace();
+					}
+
+					try {
+						value = await it.value;
+					} catch ( e ) {
+						gen.throw( e );
+					}
+
+					if ( it.done ) break;
+				}
+
+				return value;
+			};
+		}
+
+		// Check if the property is a function.
+		// Actions always run in the current scope.
+		if ( typeof result === 'function' ) {
+			return ( ...args: unknown[] ) => {
+				setNamespace( ns );
+				try {
+					return result( ...args );
+				} finally {
+					resetNamespace();
+				}
+			};
+		}
+
+		// Check if the property is an object.
+		if ( isObject( result ) ) return proxify( result, ns );
+
+		return result;
+	},
+};
+
+/**
+ * @typedef StoreProps Properties object passed to `store`.
+ * @property {Object} state State to be added to the global store. All the
+ *                          properties included here become reactive.
+ */
+
+/**
+ * @typedef StoreOptions Options object.
+ */
+
+/**
+ * Extends the Interactivity API global store with the passed properties.
+ *
+ * These props typically consist of `state`, which is reactive, and other
+ * properties like `selectors`, `actions`, `effects`, etc. which can store
+ * callbacks and derived state. These props can then be referenced by any
+ * directive to make the HTML interactive.
+ *
+ * @example
+ * ```js
+ *  store({
+ *    state: {
+ *      counter: { value: 0 },
+ *    },
+ *    actions: {
+ *      counter: {
+ *        increment: ({ state }) => {
+ *          state.counter.value += 1;
+ *        },
+ *      },
+ *    },
+ *  });
+ * ```
+ *
+ * The code from the example above allows blocks to subscribe and interact with
+ * the store by using directives in the HTML, e.g.:
+ *
+ * ```html
+ * <div data-wp-interactive>
+ *   <button
+ *     data-wp-text="state.counter.value"
+ *     data-wp-on--click="actions.counter.increment"
+ *   >
+ *     0
+ *   </button>
+ * </div>
+ * ```
+ *
+ * @param {StoreProps}   properties Properties to be added to the global store.
+ * @param {StoreOptions} [options]  Options passed to the `store` call.
+ */
+
+type DeepPartial< T > = T extends object
+	? { [ P in keyof T ]?: DeepPartial< T[ P ] > }
+	: T;
+
+interface StoreOptions {
+	lock?: boolean | string;
+}
+
+const universalUnlock =
+	'I know using a private store means my plugin will inevitably break on the next store release.';
+
+export function store< S extends object = {} >(
+	namespace: string,
+	storePart?: DeepPartial< S >,
```

Comment:
Should we add this here or should we let people use it themselves? 🤔

Reply:
Uhm, I just spent a short time thinking about this, and I'm not sure. It is better if we help developers not deal with that.

The issue is whether we want devs to get types inferred when they pass a _store part_ to `store()`. E.g., let's consider the following store:

```ts
interface MyStore {
  state: {
    prop1: string;
    prop2: string;
  }
}
```

It would be ideal if we have the arguments typed in the following call automatically:

```ts
const { state } = store< MyStore >( 'myStore', {
  state: {
    prop1: 123, // ERROR: This cannot be a number!
    prop3: 456, // ERROR: This prop does not exist!
  },
});
```

If we don't handle that internally, developers would have to do something like this:

```ts
const { state } = store< MyStore >(
  'myStore',
  {
    state: { prop1: '123' },
  } as DeepPartial< MyStore > // Just to infer types correctly.
);
```

Or perhaps they could write their own typed objects...


```ts
interface StorePart {
  state: {
    prop1: string;
  }
}

const storePart: StorePart  = {
  state: { prop1: '123' },
};

const { state } = store< MyStore >('myStore', storePart);
```

Do you have any other idea in mind, @luisherranz?

Reply:
My idea was to encourage the use of separate types and different calls to the store, one with the definition with the partial types, and one with the `state` and `actions` destructuring with the whole types.

It's a bit more complex, but the types are perfect. I explained it here in the `Objects returned by store() are typed
` section:

- https://github.com/WordPress/gutenberg/discussions/53586#:~:text=Objects%20returned%20by%20store()%20are%20typed

_By perfect, I mean that if some store props are missing, TS will complain._

My fear is that by adding `DeepPartial` inside the types of `store`, there's no way to make the types "perfect".

Reply:
I'm not 100% sure that's the ideal scenario, though. We need to do more testing.

Reply:
I think it's safer for now, yes 🙂

Reply:
Removed in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/2750a072597dc434ca28199330b6c7ce4d22a6ef.